File: go/e2e/health_check_test.go
//go:build e2e

package e2e

import (
	"context"
	"testing"

	goystore "github.com/goy-co/goy-store/go"
)

func TestHealthCheck_BackendsHealthy(t *testing.T) {
	store := createTestStore(t)
	ctx := context.Background()

	// 1. Individual contracts
	kvH, err := store.KV().IsHealthy(ctx)
	if err != nil || kvH.State != goystore.HealthHealthy || kvH.Backend != "redis" {
		t.Fatalf("KV health failed: %+v, err=%v", kvH, err)
	}

	relH, err := store.Relational().IsHealthy(ctx)
	if err != nil || relH.State != goystore.HealthHealthy || relH.Backend != "postgres" {
		t.Fatalf("Relational health failed: %+v, err=%v", relH, err)
	}

	ssH, err := store.SortedSet().IsHealthy(ctx)
	if err != nil || ssH.State != goystore.HealthHealthy || ssH.Backend != "redis" {
		t.Fatalf("SortedSet health failed: %+v, err=%v", ssH, err)
	}

	psH, err := store.PubSub().IsHealthy(ctx)
	if err != nil || psH.State != goystore.HealthHealthy || psH.Backend != "redis" {
		t.Fatalf("PubSub health failed: %+v, err=%v", psH, err)
	}

	blobH, err := store.Blob().IsHealthy(ctx)
	if err != nil || blobH.State != goystore.HealthHealthy || blobH.Backend != "local" {
		t.Fatalf("Blob health failed: %+v, err=%v", blobH, err)
	}

	// 2. Consolidated health
	health := store.HealthCheck(ctx)
	if health.State != goystore.HealthHealthy {
		t.Fatalf("Consolidated health expected healthy, got %s", health.State)
	}
	if len(health.Contracts) != 5 {
		t.Fatalf("expected 5 contracts, got %d", len(health.Contracts))
	}
}

File: rust/tests/health_check_test.rs
use super::common::*;
use goy_store::health::HealthState;

#[tokio::test]
async fn test_health_check_backends_healthy() {
    let store = create_test_store().await.expect("failed to create store");

    // 1. Check individual contracts
    let kv_health = store.kv.is_healthy().await.unwrap();
    assert_eq!(kv_health.state, HealthState::Healthy);
    assert_eq!(kv_health.contract, "kv");
    assert_eq!(kv_health.backend, "redis");

    let rel_health = store.relational.is_healthy().await.unwrap();
    assert_eq!(rel_health.state, HealthState::Healthy);
    assert_eq!(rel_health.contract, "relational");
    assert_eq!(rel_health.backend, "postgres");

    let ss_health = store.sorted_set.is_healthy().await.unwrap();
    assert_eq!(ss_health.state, HealthState::Healthy);
    assert_eq!(ss_health.backend, "redis");

    let ps_health = store.pubsub.is_healthy().await.unwrap();
    assert_eq!(ps_health.state, HealthState::Healthy);
    assert_eq!(ps_health.backend, "redis");

    let blob_health = store.blob.is_healthy().await.unwrap();
    assert_eq!(blob_health.state, HealthState::Healthy);
    assert_eq!(blob_health.backend, "local");

    // 2. Consolidated health
    let consolidated = store.health_check().await;
    assert_eq!(consolidated.state, HealthState::Healthy);
    assert_eq!(consolidated.contracts.len(), 5);
}
